### PR TITLE
Form object refactoring

### PR DESCRIPTION
**Pull request by volha-pivavarchyk**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Description:

This PR replaces [PR #11234](https://github.com/mautic/mautic/pull/11234): rebased and errors fixes

#### Steps to test this PR:

1. Open [mautic/mautic on Gitpod](https://gitpod.io/#https://github.com/mautic/mautic) or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester)) - as you need to do some things before applying the PR.
1. Before you checkout this PR:
2. Create a form with at least 4 fields. One mapped to the contact email field, one mapped to the company field under the contact opt group, one mapped to the primary company city field. The fourth field should not be mapped to any field.
3. Check out this PR - in Gitpod on the left hand side, look for the Github icon, then find this PR and right click to checkout the PR:
screenshot-mautic-mautic-oq31hzs9992 ws-eu85 gitpod io-2023 02 04-15_36_18

5. Run the migration using the terminal tab. 

Run `ddev exec bin/console doctrine:migrations:execute --up 20200415135706`. Note:  you can roll back your changes with `ddev exec bin/console doctrine:migrations:execute --down 20200415135706` if needed.

6. Open PHPMyAdmin under the Ports tab on Gitpod, look for the port 8036. Open it in a tab.

7. Open the database named 'db' and go to the `form_fields` table and check that the last 2 columns are `mapped_object` and `mapped_field`. The field that was mapped to the contact email field should have `mapped_object = 'lead'` and `mapped_field = 'email'`. The one that was mapped to the company field under the contact opt group should have `mapped_object = 'lead'` and `mapped_field = 'company'`. The one that was mapped to the primary company city field should have `mapped_object = 'company'` and `mapped_field = 'companycity'`. The form field that was not mapped at all should have null values there.
8. When you map one field then it should be removed from mapping of other fields. Example: Map contact email field to a form field. Create another field. The contact email field is not in the mapping list.
9. Test submissions of different field types.
10. Test conditional fields
11. Test progressive profiling (behaviours)

#### List deprecations along with the new alternative:
1. `form_fields.lead_field` column is deprecated. Use `mapped_object` and `mapped_field` instead.
2. `Mautic\FormBundle\Entity\Field::getLeadField()` and `setLeadField()`. Use getters and setters on `mappedObject` and `mappedField` properties instead.
3. `Mautic\FormBundle\Event\Service\FieldValueTransformer::isIsTransformed()` is deprecated as not used.

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11905"></a>

### CODE REVIEW

File: app/bundles/FormBundle/Controller/FormController.php [hunk 1]
```diff
@@ -815,27 +836,35 @@ function ($a, $b) {
...
-        \assert($leadFieldModel instanceof FieldModel);
+//        $leadFieldModel = $this->getModel('lead.field');
+//        \assert($leadFieldModel instanceof FieldModel);
+//
+//        $mappedFields  = [];
+//        $mappedObjects = $entity->getMappedFieldObjects();
+//
+//        foreach ($mappedObjects as $mappedObject) {
+//            $mappedFields[$mappedObject] = $this->fieldCollector->getFields($mappedObject);
+//        }
```
Comment (escopecz) [inline app/bundles/FormBundle/Controller/FormController.php]: Wasn't this used for something?

Comment (volha-pivavarchyk) [reply app/bundles/FormBundle/Controller/FormController.php]: I think it is unnecessary code. `contactFields` and  `companyFields` are not used, so neither is `$leadFieldModel`. `mappedFields` and `mappedObjects` are not used anywhere either.

**Review by RCheesley — DISMISSED** [review]
Thanks for picking this up @volha-pivavarchyk - works wonderfully!  Fields are correctly mapped once the migration is run:

screenshot-8036-mautic-mautic-oq31hzs9992 ws-eu85 gitpod io-2023 02 04-15_13_13

screenshot-8080-mautic-mautic-oq31hzs9992 ws-eu85 gitpod io-2023 02 04-15_14_40

Note, called Contact not Lead!

I tested submitting forms with all the different field types, including conditional fields and everything was saved as I would have expected. 🚀

**Review by mabumusa1 — DISMISSED** [review]
I tested locally, this PR is a great addition. It works as described

**Comment by RCheesley** [comment]
RTC assuming the comments from @escopecz on the code review are satisfied, which it looks like they are.

File: app/bundles/FormBundle/Controller/FormController.php (hunk 1 above)
Comment (volha-pivavarchyk) [reply app/bundles/FormBundle/Controller/FormController.php]: @escopecz I think this code should be removed. WDYT?

Comment (volha-pivavarchyk) [reply app/bundles/FormBundle/Controller/FormController.php]: I removed the comented code.

**Comment by volha-pivavarchyk** [comment]
@RCheesley @escopecz @mabumusa1 I removed the commented code that was unnecessary. This code is not used and shouldn't affect anything. Could you please test and re-approve it again.

File: app/bundles/FormBundle/Controller/FormController.php (hunk 1 above)
Comment (mabumusa1) [reply app/bundles/FormBundle/Controller/FormController.php]: I retested this PR, and the removal of this code, did not impact the functionality of the code. so I think this conversation should be resolved and this PR be merged

**Review by mabumusa1 — APPROVED** [review]
Approving this PR again, after the removal of unused code

**Comment by RCheesley** [comment]
Thanks for all the testing folks, will commit this once the tests complete! 🚀

**Pull request merged**